### PR TITLE
PD: Extend InvoluteGear to make the tooth length adjustable

### PR DESCRIPTION
**Pull request by jbaehr**
This PR exposes the addendum (tooth height above pitch circle), dedendum (tooth height from its root to the pitch circle) and root fillet radius as properties of PartDesign's Involute Gear profile. The values are all exposed as coefficients, i.e. normalized by the gear's module.

This PR addresses the following use cases:
- Adjust the tooth hight for internal gears to avoid interferences or non-involute flanks
- Create gears using others than the "full-deptsh system", e.g. with [stub teeth](https://khkgears.net/new/gear_knowledge/gear-nomenclature/stub-teeth.html)
- Use the profile for [involute splines](https://en.wikipedia.org/wiki/Spline_(mechanical)), i.e. splined shafts and hubs

image

When loading documents with gears created in earlier versions, the new properties are setup with values matching the previously hard-coded ones. This way the generated profile does not change. Unit tests for the backward compatibility are provided with fixtures generated by official mac build of FreeCAD-0.20.1.

For review, I suggest to look at the commits individually. They are carefully crafted to address a specific concern each and do not mix refactors/cleanups with functional changes. For this reason, please *do not squash* them at merge.

- [x]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8184">pipeline status</a> for feature branch [PR_8184](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8184). Pipeline [747073365](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747073365) was triggered at [edea040](https://github.com/FreeCAD/FreeCAD/pull/8184/commits/edea040f82c168be0b440e2f79912b942be1446c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/PartDesign/fcgear/involute.py
Comment (luzpaz) [inline src/Mod/PartDesign/fcgear/involute.py:192]: pls fix typo truely ==> truly

File: src/Mod/PartDesign/InvoluteGearFeature.py
Comment (luzpaz) [inline src/Mod/PartDesign/InvoluteGearFeature.py:127]: This is not related to your PR but do you mind adding a separate commit to remove this unnecessary semicolon?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8184">pipeline status</a> for feature branch [PR_8184](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8184). Pipeline [747309730](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747309730) was triggered at [4e753e2](https://github.com/FreeCAD/FreeCAD/pull/8184/commits/4e753e2dcf43feaab96f646f4d70a8d0e27244f7). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/PartDesign/CMakeLists.txt
Comment (donovaly) [inline src/Mod/PartDesign/CMakeLists.txt:59]: Can you please use a generic name. These files will stay also for FC 0.21 or 1.0

File: src/Mod/PartDesign/InvoluteGearFeature.ui
Comment (donovaly) [inline src/Mod/PartDesign/InvoluteGearFeature.ui:199]: Please also add tooltips. These will help especially newbies a lot.

**Review by donovaly — CHANGES_REQUESTED** [review]
Many thanks. Looks good.
I only have these easy to resolve remarks:
- since I cannot test this right now, have you tested that files created with FreeCAD 0.20 will work fine (no warnings) also with your changes?
- please use generic file names without referencing a FreeCAD version
- please add tooltips to the UI file
- can you please merge your different commits to one

File: src/Mod/PartDesign/InvoluteGearFeature.ui
Comment (jbaehr) [reply src/Mod/PartDesign/InvoluteGearFeature.ui:199]: Yes, that's on my TODO list. My plan is to use the property descriptions as tool tips, in order to not maintain two times the same string. And make them translatable. Basically the same as what I've done for the Helix, cf. #5321.
I thought it would better be done as a separate PR -- separation if concerns, you know ;-)

File: src/Mod/PartDesign/CMakeLists.txt
Comment (jbaehr) [reply src/Mod/PartDesign/CMakeLists.txt:59]: The name is on purpose. Those files are used by unit tests to ensure we can 
a) load the old files
b) recalculate them successfully
b) come to the same result as with earlier versions. 

So if I have a unit test, asserting compat with v0.20.1, I do not want a generic name, but a name that makes clear that those are files from v0.20.1

**Comment by jbaehr** [comment]
Thanks for the review.

> * since I cannot test this right now, have you tested that files created with FreeCAD 0.20 will work fine (no warnings) also with your changes?

Yes, as stated in he PR's description. I also added unit tests for this scenario to backup my claim and avoid future regressions. Since the InvoluteGear has not seen any changes lately, I guess using only v0.20.1 as reference is enough, though.

> * please use generic file names without referencing a FreeCAD version

Why? Those files are explicitly from v0.20.1, to ensure compat with that very version. If I add new features after we have released a, lets say 1.0, and I add new features for a potential 1.1, I will add additional fixtures generated with 1.0 which will then have "v1-0-0" in their name. And once we decide that backward compat with v0.20 is not needed any more, we can remove the "v0-20-1" files.

> * please add tooltips to the UI file

Will do, but on a dedicated PR, together with translatability and single-source for property docs and task panel tool tips. As there are no tool tips for all the previous fields as well, I decided to group my PRs that way.
Is that OK for you?

> * can you please merge your different commits to one

Here I oppose. I have carefully crafted my commits to address one specific concern each. I explicitly did not mix refactors/cleanups with functional changes. And I stated the motivation behind each change in the respective commit messages.  
I see a benefit of *not squashing everything together*. Not only to ease your review, but also to enable sensible bisection in case of regressions or make efficient use of git's history (e.g. via blame/annotate) in the future.
To not repeat what others have already put together very nicely, please see [this excellent article by Chris Beams](https://cbea.ms/git-commit/) with which I couldn't agree more.

File: src/Mod/PartDesign/InvoluteGearFeature.ui
Comment (donovaly) [reply src/Mod/PartDesign/InvoluteGearFeature.ui:199]: Fine with me.

File: src/Mod/PartDesign/CMakeLists.txt
Comment (donovaly) [reply src/Mod/PartDesign/CMakeLists.txt:59]: Well, only experts will understand this. Especially the reference to 0.20.1 (the ".1") is a problem. Here you add a new feature, so it will not be backported. Therefore, if you insist to include the version name, please use then only 0.20.

**Comment by donovaly** [comment]
> I see a benefit of not squashing everything together.

Especially when running a Git blame this is a problem. It is often to find the PR that introduced the problem, rather than the particular commit This saves time for maintainers like me and usually the PR submitter has then already enought knowledge to check his PR.
Therefore we usually keep the commits per PR low. If we think having several commits is more suitable, we make several PRs. Sure this is a matter of taste, as a reviewer I only follow here what we usually do. If you insist to have the different commits, then I request you to split the PR to one PR in which are the commits where you add the actual feature and one where you refactor the code. Would that be OK for you?

**Comment by jbaehr** [comment]
> > I see a benefit of not squashing everything together.
> 
> Especially when running a Git blame this is a problem. It is often to find the PR that introduced the problem, rather than the particular commit This saves time for maintainers like me and usually the PR submitter has then already enought knowledge to check his PR.

Well, then consider it that way: If you come back to me later, because my PR caused a problem, I definitely prefer digging though a few smaller well described  changes, then a single big one. I strive to write the git history I'd like to read.

> Therefore we usually keep the commits per PR low. If we think having several commits is more suitable, we make several PRs. Sure this is a matter of taste, as a reviewer I only follow here what we usually do.

I'm totally with when you want to keep the changes per PR low. That's also a reason why I choose not to throw in the tooltip/translation stuff right now. But just replacing a few smaller commits with a single big one, doesn't affect the changes made in the PR at all. It just makes them harder to understand.

> If you insist to have the different commits, then I request you to split the PR to one PR in which are the commits where you add the actual feature and one where you refactor the code. Would that be OK for you?

Well, "insisting" is maybe a bit too strong. In my eyes it's just a pity to throw away valuable information that may help future developers (myself included) understanding the reasoning behind certain changes. If feels like I'm forced to make my PR worse, not better. That's my main reason to be reluctant.

Sure I could split it in two or three dedicated PRs, but what do you gain by this? The changes are the same, you just get additional merges in between (and thus lose a bit of context).

File: src/Mod/PartDesign/CMakeLists.txt
Comment (jbaehr) [reply src/Mod/PartDesign/CMakeLists.txt:59]: I feel we still have a misunderstanding here. This is not about backporting, it's about ensuring backward compatibility.

I wrote automated tests that ensure the current version (0.21-alpha, 1.0, what ever the future brings) is still able to correctly handle gears created with FreeCAD 0.20.1. That was my safety net during development and kinda insurance for my older projects against potential regressions in the future.

As the gear's code hasn't changed in the recent past, it shouldn't really matter whether I'd use 0.20.0, 0.20.1 or 0.20.2 -- I just happen to have a 0.20.1 around to create those reference files. And because it was a 0.20.1 I suffixed them with "_v0-20-1".

Is your proposal to recreate the reference files with the original 0.20.0 and name them "..._v0-20" instead? Or is having the patch level in file name the problem? If so, why?
Or is it still unclear what I want to achieve here?

**Comment by chennes** [comment]
Just as another data point: I prefer the current commit structure to one with any merging. I think @jbaehr did a very nice job structuring this PR, and I'd encourage other developers to follow this good example.

**Comment by donovaly** [comment]
> Sure I could split it in two or three dedicated PRs, but what do you gain by this? The changes are the same, you just get additional merges in between (and thus lose a bit of context).

Well, once I started to do some merging work, I was explicitly told to keep the number of commits per PR low, if there are additions and refactorings, ideally the refactoring should be sorted out to a separate PR. I thought this is the case here.
You had good reasons that this is not necessary and personally, I don't have a real strong opinion, just tried to act as I usually do.
So let's keep the PR as it is, except of the naming, see my next post

In general, there is no rule how to handle the commits within a PR, so there is no right or wrong.

**Comment by donovaly** [comment]
> As the gear's code hasn't changed in the recent past, it shouldn't really matter whether I'd use 0.20.0, 0.20.1 or 0.20.2 -- I just happen to have a 0.20.1 around to create those reference files. And because it was a 0.20.1 I suffixed them with "_v0-20-1".

But this is my point. the "-1" implies for other or future developers that there was a change in the 0.20 code line. But since this was not the case, the naming can be confusing (a future developer might spend time to onvestigate what might happened between 0.20.0, 0.20.1 0.20.2 etc.)

Therefore in my opinion the naming should be just "_v0-20".

> Is your proposal to recreate the reference files with the original 0.20.0 and name them "..._v0-20" instead?

Not completely. Since there were no changes in the 0.20 code line for the gear code, the name should be "..._v0-20" but there is no need to use 0.20.2 to recreate them, you can use 0.20.1 or 0.20.2.
 Or is having the patch level in file name the problem? If so, why? Or is it still unclear what I want to achieve here?

File: src/Mod/PartDesign/CMakeLists.txt
Comment (jbaehr) [reply src/Mod/PartDesign/CMakeLists.txt]: (Answering to [this comment](https://github.com/FreeCAD/FreeCAD/pull/8184#issuecomment-1399381491), which somehow got out of context)

> > As the gear's code hasn't changed in the recent past, it shouldn't really matter whether I'd use 0.20.0, 0.20.1 or 0.20.2 -- I just happen to have a 0.20.1 around to create those reference files. And because it was a 0.20.1 I suffixed them with "_v0-20-1".
> 
> But this is my point. the "-1" implies for other or future developers that there was a change in the 0.20 code line. But since this was not the case, the naming can be confusing (a future developer might spend time to onvestigate what might happened between 0.20.0, 0.20.1 0.20.2 etc.)
> 
> Therefore in my opinion the naming should be just "_v0-20".

This is a reasoning I can follow. So I kept the version number in the file name, but removed patch-level.

**Comment by donovaly** [comment]
Thanks for the naming change.

-------------------
Since there internal discussions why I requested a PR split or commit merge. Here if only for information was what I initially had in mind when I see the PR the first time:
I feel like being brought to court in this respect. But following the way I review PRs, I would have done like this:
- commit 1 is one PR since it adds a testcase that can also be backported
- commit 2 and 3 can become another PR since this is code refactoring
- while commit 4 to end add a new feature and could be a separate PR

**Comment by donovaly** [comment]
The last step is to update the Wiki. Can you therefore please add the new feature to our announcement list:
https://wiki.freecadweb.org/Release_notes_1.0
and update the Wiki pages describing the feature:
https://wiki.freecadweb.org/PartDesign_InvoluteGear

--------------
Please also make a PR with the tooltip addition.

**Comment by jbaehr** [comment]
Thank you.

I will address those three points over the upcoming week.

**Pull request merged**
